### ISSUE TITLE
`float` type parameter lead to slow convergence of `adam` 

### ISSUE THREAD

**Issue by yihengwuKP** [deepmind/optax#333]
# Summary
I found that if I initialize the parameter to `int` type, then the `adam` would find the minimum faster. Also the `sgd` would be faster than `adam` in my testing case with the same `learning_rate`. Is there anything I did wrong? 

# Code to reproduce
```python
import optax

def calc_loss(p):
    return p**2

def calc_grad(p):
    return 2*p

# float with adam
para = 100.
optimizer = optax.adam(learning_rate=1e-1)
opt_state = optimizer.init(para)
for i in range(1000):
    grad = calc_grad(para)
    updates, opt_state = optimizer.update(grad, opt_state, para)
    para = optax.apply_updates(para, updates)
    if i % 100 == 0:
        print(f"step {i}, loss: {calc_loss(para)}")
```
step 0, loss: 9980.0107421875
step 100, loss: 8114.1650390625
step 200, loss: 6521.810546875
step 300, loss: 5178.337890625
step 400, loss: 4057.343505859375
step 500, loss: 3133.39599609375
step 600, loss: 2382.150390625
step 700, loss: 1780.4334716796875
step 800, loss: 1306.3935546875
step 900, loss: 939.6651000976562

But if changed to `para=100`, the output would be:
step 0, loss: 9801
step 100, loss: 0
step 200, loss: 0
step 300, loss: 0
step 400, loss: 0
step 500, loss: 0
step 600, loss: 0
step 700, loss: 0
step 800, loss: 0
step 900, loss: 0

If I changed to sgd with float `para=100.`, then the output is:
step 0, loss: 6400.0
step 100, loss: 2.655692076312363e-16
step 200, loss: 1.1019828342605891e-35
step 300, loss: 0.0
step 400, loss: 0.0
step 500, loss: 0.0
step 600, loss: 0.0
step 700, loss: 0.0
step 800, loss: 0.0
step 900, loss: 0.0

Are these behaviours intended? Thank you!

**Comment by yihengwuKP** [comment]
I am using optax `0.1.1`, jax `0.3.1`

**Comment by rosshemsley** [comment]
Hey @yihengwuKP,

Adding in some prints, we can see that the updates are small negative values, e.g. `-0.09999934`.
When we add these updates to the params, and then cast (which is what apply_updates does), these values get truncated **down** to the nearest integer.

This means that the parameter decreases by a whole integer step on every gradient step (99, 98, 97, ...).

For most optimizations, this might be disastrous - usually the optimal value is not an integer. However, since the objective function here is x^2 - the true optimum sits at x==0, achieving a loss of 0.

This means the optimizer quickly finds the optimal value and converges.

For the floating point case, the optimizer proceeds much more slowly (since there's no integer truncation). For this very specific optimization case, it turns out that this is slower. In general though, this would not be the case :) 


Does that help?

**Comment by yihengwuKP** [comment]
wooo, many thanks for the detailed explanation, it makes perfect sense now!
